src/field/gen_fieldops.py

```diff
@@ -0,0 +1,251 @@
+#!/usr/bin/env python3
+"""Code-generator for arithmetic operators on Field2Ds/Field3Ds
+
+This uses the jinja template in gen_fieldops.jinja to generate code
+for the arithmetic operators, and prints to stdout.
+
+The `Field` class provides some helper functions for determining how to
+pass a variable by reference or pointer, and how to name arguments in
+function signatures. This allows us to push some logic into the
+templates themselves.
+
+"""
+
+
+from __future__ import print_function
+
+try:
+    from builtins import object
+except ImportError:
+    pass
+
+import argparse
+from collections import OrderedDict
+import contextlib
+from copy import deepcopy as copy
+import itertools
+import sys
+
+try:
+    import jinja2
+except ImportError:
+    raise ImportError('Missing Python module "jinja2". See "Field2D/Field3D Arithmetic '
+                      'Operators" in the BOUT++ user manual for more information')
+
+
+# This allows us to open either a file or stdout with the same code
+@contextlib.contextmanager
+def smart_open(filename, mode='r'):
+    """Open stdin or stdout using a contextmanager
+
+    From: http://stackoverflow.com/a/29824059/2043465
+    """
+    if filename == '-':
+        if mode is None or mode == '' or 'r' in mode:
+            fh = sys.stdin
+        else:
+            fh = sys.stdout
+    else:
+        fh = open(filename, mode)
+    try:
+        yield fh
+    finally:
+        if filename is not '-':
+            fh.close()
+
+
+# The arthimetic operators
+# OrderedDict to (try to) ensure consistency between python 2 & 3
+operators = OrderedDict([
+    ('*', 'multiplication'),
+    ('/', 'division'),
+    ('+', 'addition'),
+    ('-', 'subtraction'),
+])
+
+header = """// This file is autogenerated - see gen_fieldops.py
+#include <bout/mesh.hxx>
```

Comment:
`utils.hxx` is required for `checkData` on `BoutReal` (otherwise can get ambiguous errors as tries to cast to either 2d/3d if these versions are available). I assume something you're including is including this for us at the moment as this obviously compiles fine as it passes the tests.

Reply:
Ah, this is true. I'm going to merge this now, but I'll revisit this when we start using Block-Region/SDI